semantics/src/main/kotlin/com/strumenta/kolasu/semantics/symbol/provider/declarative/DeclarativeSymbolProvider.kt

```diff
@@ -0,0 +1,165 @@
+package com.strumenta.kolasu.semantics.symbol.provider.declarative
+
+import com.strumenta.kolasu.ids.NodeIdProvider
+import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.model.ReferenceByName
+import com.strumenta.kolasu.semantics.symbol.description.BooleanValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.ContainmentValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.IntegerValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.ListValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.NullValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.ReferenceValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.StringValueDescription
+import com.strumenta.kolasu.semantics.symbol.description.SymbolDescription
+import com.strumenta.kolasu.semantics.symbol.description.ValueDescription
+import com.strumenta.kolasu.semantics.symbol.provider.SymbolProvider
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.allSuperclasses
+import kotlin.reflect.full.isSuperclassOf
+import kotlin.reflect.full.safeCast
+
+inline fun <reified NodeTy : Node> symbolFor(
+    noinline specification: DeclarativeSymbolProvideRuleApi<NodeTy>.(
+        DeclarativeSymbolProviderRuleContext<NodeTy>
+    ) -> Unit
+): DeclarativeSymbolProviderRule<NodeTy> {
+    return DeclarativeSymbolProviderRule(NodeTy::class, specification)
+}
+
+open class DeclarativeSymbolProvider(
+    private val nodeIdProvider: NodeIdProvider,
+    vararg rules: DeclarativeSymbolProviderRule<out Node>
+) : SymbolProvider {
+    private val rules: List<DeclarativeSymbolProviderRule<out Node>> = rules.sorted()
+
+    override fun symbolFor(node: Node): SymbolDescription {
+        return this.rules
+            .firstOrNull { it.isCompatibleWith(node::class) }
+            ?.invoke(nodeIdProvider, this, node)
+            ?: throw RuntimeException("No compatible rule found for ${node::class.qualifiedName}")
```

Comment:
What is not clear to me is what I should do to declare symbols only for some nodes. Before I think I was just not providing a rule in the DeclarativeSymbolProvider, but now it seems that I must always provide a symbol for a node. I thought that the idea was to provide symbols only for nodes that we want to make visibile from outside the AST

Reply:
You're right, I will restore the previous interface.